### DISCUSSION TITLE
How to structure the project with Kratos?

### DISCUSSION THREAD

**Discussion by spa5k** [ory/kratos#1810] [Q&A] [answered]
## Let's say, I have a typical setup, a next js website for the frontend, a separate nodejs based backend using Postgres for data.

So, now I want to add Kratos to this setup, how should i work on adding it optimally?

 **1. Adding kratos on same database as the one getting used by the main backend.**

	Pro - no separate database management.
	Cons - it mixes up kratos tables onto main schema.

 **2. Adding kratos on a different database but on same docker container as the one used by main backend.**

	Pro - not much of a separate database management.
	Does not add tables and things to main database.

So what will be the optimal way to access data from kratos in my project.

**For first scenario -** 

1. Since kratos is in same db I can just reference `identity id` to a `kratos id` in users table and use it for reference.

**For second scenario-** 
1. Keep a column `kratos id` in my users table, and then query to the kratos database, probably through the http interface or something, idk.

**Comment by vinckr** [comment, marked as answer]
It seems like the second scenario has fewer drawbacks. 
As you have more practical experience with this setup, would you agree @Benehiko ?

**Reply by Benehiko** [reply]
Hi @Unbuttun 

I would keep the database separate (but you can keep it on the same PostgreSQL instance) between Kratos and your actual application. I would then add an identity UUID column in my application which is basically the identity UUID given by kratos. 

You have a couple of options to then populate the columns for a user upon user registration (a new user is created).
- Either you add a middleware that checks the cookie against Kratos with a `/sessions/whoami` and retrieves all the information and then compare it against your DB (does this user exist?)
- Have an endpoint on your application that can be called by Kratos post-registration hook. See the [web-hooks docs](https://www.ory.sh/kratos/docs/self-service/hooks#web-hooks)

To get the user ID on a request just call the Kratos `/sessions/whoami` endpoint from your application (middleware).
See the [HTTP API docs](https://www.ory.sh/kratos/docs/reference/api#operation/toSession)

**Reply by spa5k** [reply]
Thanks for the help @Benehiko. 🥂
